libbeat/common/url.go

```diff
@@ -83,3 +83,37 @@ func EncodeURLParams(url string, params url.Values) string {
 
 	return strings.Join([]string{url, "?", params.Encode()}, "")
 }
+
+type ParseHint func(raw string) string
+
+// ParseURL tries to parse a URL and return the parsed result.
+func ParseURL(raw string, hints ...ParseHint) (*url.URL, error) {
+	if raw == "" {
+		return nil, nil
+	}
+
+	u, err := url.Parse(raw)
+	if err == nil && strings.HasPrefix(u.Scheme, "http") {
```

Comment:
This still requires 'http' as default scheme. We might even want to support multiple schemes like `http` and `https` for this URL. I wonder if we should even check the exact scheme. Maybe just check for `u.Scheme != ""`.

Reply:
Well, it is using `strings.HasPrefix` so both `http` and `https` would match. But, as you say, we might as well just check for `u.Scheme != ""` at that point. I've made this change in c5fde6ff3be765a89c0bc20f9cae8f697d08d47e.

Reply:
The change I made is causing tests to fail like so:

```
go test github.com/elastic/beats/v7/libbeat/common                                                                                                  1 ↵
--- FAIL: TestParseURL (0.00s)
    --- FAIL: TestParseURL/no_scheme (0.00s)
        url_test.go:163:
                Error Trace:    url_test.go:163
                Error:          Not equal:
                                expected: "http://host:1234/path"
                                actual  : "host:1234/path"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -http://host:1234/path
                                +host:1234/path
                Test:           TestParseURL/no_scheme
    --- FAIL: TestParseURL/default_scheme_https (0.00s)
        url_test.go:163:
                Error Trace:    url_test.go:163
                Error:          Not equal:
                                expected: "https://host:1234/path"
                                actual  : "host:1234/path"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -https://host:1234/path
                                +host:1234/path
                Test:           TestParseURL/default_scheme_https
    --- FAIL: TestParseURL/invalid (0.00s)
        url_test.go:161:
                Error Trace:    url_test.go:161
                Error:          An error is expected but got nil.
                Test:           TestParseURL/invalid
FAIL
FAIL    github.com/elastic/beats/v7/libbeat/common      0.245s
FAIL
```

Essentially, it looks like `url.Parse` will not return an error when given a URL without any scheme, so we do need to add the extra condition for the scheme starting with `http`.

So I'm reverting the change I made in c5fde6f.

Reply:
yeah, url.Parse only works well if you give it a correct URL :/

A recent community PR introduced this: https://github.com/elastic/beats/blob/master/libbeat/outputs/redis/redis.go#L120

One could replace the SplitN with a call to Find.

Reply:
I implemented the same logic as shown in the community PR in 023409245. 

> One could replace the SplitN with a call to Find.

Which `Find` are you referring to? I could not find that function in the `strings` package.

Reply:
Sorry, wrong language. I meant strings.Index.

Reply:
Changed in 9de9b87e2ca5593e68ee86547d7f9c7f122ac10c.